onnxruntime/core/common/threadpool.cc

```diff
@@ -23,114 +22,202 @@ limitations under the License.
+    // most/all of the thread pool's threads are active in the loop.  Threads outside the pool may
+    // also be claiming work, with CurrentThreadId -1.
+    int num_threads = _tp.NumThreads();
+    int my_thread_idx = (_tp.CurrentThreadId() + 1) % num_threads;
+    assert(my_thread_idx >= 0 && my_thread_idx < num_threads);
+
+    int home_shard;
+    if (num_threads >= NUM_SHARDS) {
+      // More threads than shards => allocate them home shards round-robin, aiming to sprace the load across
+      // the shards
+      home_shard = my_thread_idx % NUM_SHARDS;
+    } else {
+      // Fewer threads than shards => spread the threads evenly across the shards, so each will work
+      // on a run of successive shards before contention
+      home_shard = (my_thread_idx * NUM_SHARDS) / num_threads;
     }
+    assert(home_shard >= 0 && home_shard < NUM_SHARDS);
+    return home_shard;
   }
-  // Wait for the specified time, return false iff the count has not dropped to
-  // zero before the timeout expired.
-  inline bool WaitFor(std::chrono::milliseconds ms) {
-    unsigned int v = state_.fetch_or(1, std::memory_order_acq_rel);
-    if ((v >> 1) == 0)
-      return true;
-    std::unique_lock<OrtMutex> l(mu_);
-    while (!notified_) {
-      const std::cv_status status = cond_var_.wait_for(l, ms);
-      if (status == std::cv_status::timeout) {
-        return false;
+
+  // Attempt to claim iterations from the sharded counter.  The function either
+  // returns true, along with a block of exactly block_size iterations, or it returns false
+  // if all of the iterations have been claimed.
+  bool ClaimIterations(int my_home_shard,
+                       int& my_shard,
+                       uint64_t& my_start,
+                       uint64_t& my_end) {
+    do {
+      if (_shards[my_shard]._next < _shards[my_shard]._end) {
+        // Appears to be work in the current shard, try to claim with atomic fetch-and-add
+        uint64_t temp_start = _shards[my_shard]._next.fetch_add(_block_size);
+        if (temp_start < _shards[my_shard]._end) {
+          my_start = temp_start;
+          my_end = std::min(_shards[my_shard]._end, temp_start + _block_size);
+          return true;
+        }
       }
-    }
-    return true;
+      // Work in the current shard is exhausted, move to the next shard, until
+      // we are back at the home shard.
+      my_shard = (my_shard + 1) % NUM_SHARDS;
+    } while (my_shard != my_home_shard);
+    return false;
   }
 
- private:
-  OrtMutex mu_;
-  OrtCondVar cond_var_;
-  std::atomic<int> state_;  // low bit is waiter flag
-  bool notified_;
+private:
+  alignas(CACHE_LINE_BYTES) LoopCounterShard _shards[NUM_SHARDS];
+  const ThreadPool& _tp;
+  const uint64_t _block_size;
 };
-}  // namespace
-namespace concurrency {
+
+#ifndef __GNUC__
+#pragma warning(pop) /* Padding added in LoopCounterShard, LoopCounter */
```

Comment:
Update to use #ifdef _MSC_VER

Reply:
Thanks, sorry I missed this one first time.